Wrappers/Python/cil/optimisation/functions/IndicatorBox.py

```diff
@@ -85,11 +110,125 @@ def proximal_conjugate(self, x, tau, out=None):
-            return x - tau * self.proximal(x/tau, tau)
+            # x - tau * self.proximal(x/tau, tau):
+            # use x as temporary storage variable
+            x/=tau
+            out = self.proximal(x, tau)
+            out *= -1*tau
+            # restore the values of x
+            x*=tau
+            out += x
+            return out
         
         else:
-            
             self.proximal(x/tau, tau, out=out)
             out *= -1*tau
             out += x
+
+## Utilities
+def _get_as_nparray_or_number(x):
+    '''Returns x as a numpy array or a number'''
+    try:
+        return x.as_array()
+    except AttributeError:
+        # In this case we trust that it will be either a numpy ndarray 
+        # or a number as described in the docstring
+        return x
+
+@numba.jit(nopython=True)
+def _array_within_limits_ff(x, lower, upper):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    go_ahead = True
+    for i in numba.prange(x.size):
+        if go_ahead and (x.flat[i] < lower or x.flat[i] > upper):
+            go_ahead = False
+            break
+    return 0 if go_ahead else np.inf
+
+@numba.jit(nopython=True)
+def _array_within_limits_af(x, lower, upper):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    if x.size != lower.size:
+        raise ValueError('x and lower must have the same size')
+    go_ahead = True
+    for i in numba.prange(x.size):
+        if go_ahead and (x.flat[i] < lower.flat[i] or x.flat[i] > upper):
+            go_ahead = False
+            break
+    return 0 if go_ahead else np.inf
+
+@numba.jit(nopython=True)
+def _array_within_limits_aa(x, lower, upper):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    if x.size != lower.size or x.size != upper.size:
+        raise ValueError('x, lower and upper must have the same size')
+    go_ahead = True
+    for i in numba.prange(x.size):
+        if go_ahead and (x.flat[i] < lower.flat[i] or x.flat[i] > upper.flat[i]):
+            go_ahead = False
+            break
+    return 0 if go_ahead else np.inf
+
+@numba.jit(nopython=True)
+def _array_within_limits_fa(x, lower, upper):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    if x.size != upper.size:
+        raise ValueError('x and upper must have the same size')
+    go_ahead = True
+    for i in numba.prange(x.size):
+        if go_ahead and (x.flat[i] < lower or x.flat[i] > upper.flat[i]):
+            go_ahead = False
+            break
+    return 0 if go_ahead else np.inf
+##########################################################################
+@numba.jit(nopython=True)
+def _proximal_aa(x, lower, upper, out):
+    '''Similar to np.clip except that the clipping range can be defined by ndarrays'''
+    if x.size != lower.size or x.size != upper.size:
+        raise ValueError('x, lower and upper must have the same size')
+    for i in numba.prange(x.size):
+        if x.flat[i] < lower.flat[i]:
```

Comment:
can this be an if, elif, else as only one can be true.

Reply:
I don't understand: they both need to be equal in size to x, so either failing to do so will raise.